pylint/extensions/typing.py

```diff
@@ -328,6 +367,57 @@ def _check_broken_noreturn(self, node: Union[nodes.Name, nodes.Attribute]) -> No
                 self.add_message("broken-noreturn", node=node, confidence=INFERENCE)
                 break
 
+    def _check_broken_callable(self, node: Union[nodes.Name, nodes.Attribute]) -> None:
+        """Check for 'collections.abc.Callable' inside Optional and Union."""
+        inferred = safe_infer(node)
+        if not (
+            isinstance(inferred, nodes.ClassDef)
+            and inferred.qname() == "_collections_abc.Callable"
+            and self._broken_callable_location(node)
+        ):
+            return
+
+        self.add_message("broken-collections-callable", node=node, confidence=INFERENCE)
+
+    def _broken_callable_location(  # pylint: disable=no-self-use
+        self, node: Union[nodes.Name, nodes.Attribute]
+    ) -> bool:
+        """Check if node would be a broken location for collections.abc.Callable."""
+        if is_postponed_evaluation_enabled(node) and is_node_in_type_annotation_context(
+            node
+        ):
+            return False
+
+        # Check first Callable arg is a list of arguments -> Callable[[int], None]
+        if not (
+            isinstance(node.parent, nodes.Subscript)
+            and isinstance(node.parent.slice, nodes.Tuple)
+            and len(node.parent.slice.elts) == 2
+            and isinstance(node.parent.slice.elts[0], nodes.List)
+        ):
+            return False
+
+        # Check nested inside Optional or Union
+        parent_subscript = node.parent.parent
+        if isinstance(parent_subscript, nodes.BaseContainer):
+            parent_subscript = parent_subscript.parent
+        if not (
+            isinstance(parent_subscript, nodes.Subscript)
+            and isinstance(parent_subscript.value, (nodes.Name, nodes.Attribute))
+        ):
+            return False
```

Comment:
This line is not covered by the tests.

Reply:
Saw that too. Seems like my previous filter steps are quite good already. Took me a moment to find something that would run this line.